x/airdrop/keeper/claim.go

```diff
@@ -0,0 +1,195 @@
+	errorsmod "cosmossdk.io/errors"
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/Stride-Labs/stride/v22/x/airdrop/types"
+)
+
+// User transaction to claim all the pending airdrop rewards up to the current day
+func (k Keeper) ClaimDaily(ctx sdk.Context, airdropId, claimer string) error {
+	// Fetch the airdrop and user's allocations
+	airdrop, airdropFound := k.GetAirdrop(ctx, airdropId)
+	if !airdropFound {
+		return types.ErrAirdropNotFound.Wrapf("airdrop %s", airdropId)
+	}
+	userAllocation, userFound := k.GetUserAllocation(ctx, airdropId, claimer)
+	if !userFound {
+		return types.ErrUserAllocationNotFound.Wrapf("user %s for airdrop %s", claimer, airdropId)
+	}
+
+	// Confirm the user has not elected the non-daily claim types
+	if userAllocation.ClaimType != types.CLAIM_DAILY {
+		return types.ErrClaimTypeUnavailable.Wrapf("user has already elected claim option %s",
+			userAllocation.ClaimType.String())
+	}
+
+	// Get the index in the allocations array from the current date
+	// E.g. on the 10th day of distribution, this will map to the 9th index in the list
+	todaysIndex, err := airdrop.GetCurrentDateIndex(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Sum the rewards up to that date and 0 them out in the process
+	todaysRewards := sdkmath.ZeroInt()
+	for i := 0; i <= todaysIndex; i++ {
+		rewardsOnDate := userAllocation.Allocations[i]
+		todaysRewards = todaysRewards.Add(rewardsOnDate)
+		userAllocation.Allocations[i] = sdkmath.ZeroInt()
+	}
+
+	// If there are no rewards, alert the user with an error
+	if todaysRewards.IsZero() {
+		return types.ErrNoUnclaimedRewards
+	}
+
+	// Update the amount claimed on the allocation record
+	userAllocation.Claimed = userAllocation.Claimed.Add(todaysRewards)
+
+	// Distribute rewards from the distributor
+	distributorAccount := sdk.MustAccAddressFromBech32(airdrop.DistributionAddress)
+	claimerAccount := sdk.MustAccAddressFromBech32(userAllocation.Address)
+	rewardsCoin := sdk.NewCoin(airdrop.RewardDenom, todaysRewards)
+
+	if err := k.bankKeeper.SendCoins(ctx, distributorAccount, claimerAccount, sdk.NewCoins(rewardsCoin)); err != nil {
+		return errorsmod.Wrapf(err, "unable to distribute rewards")
+	}
+
+	// Update the reward record for to mark the progress
+	k.SetUserAllocation(ctx, userAllocation)
+
+	return nil
+}
+
+// User transaction to claim a portion of their total amount now, and forfeit the
+// remainder to be clawed back
+func (k Keeper) ClaimEarly(ctx sdk.Context, airdropId, claimer string) error {
+	// Fetch the airdrop and user's allocations
+	airdrop, airdropFound := k.GetAirdrop(ctx, airdropId)
+	if !airdropFound {
+		return types.ErrAirdropNotFound.Wrapf("airdrop %s", airdropId)
+	}
+	userAllocation, userFound := k.GetUserAllocation(ctx, airdropId, claimer)
+	if !userFound {
+		return types.ErrUserAllocationNotFound.Wrapf("user %s for airdrop %s", claimer, airdropId)
+	}
+
+	// Confirm the user has not elected the non-daily claim types
```

Comment:
nit
```suggestion
	// Confirm the user has not elected the daily claim types
```

Reply:
i the original comment is correct

Reply:
ah you're right

maybe for slightly more clarity we could make it "Confirm the user has elected the daily claim types" instead of the double negative? 

defer to you, and it's a nit point. feel free to resolve

Reply:
Yeah agreed it is confusing, it's also slightly stale since we used to have claim and stake.

The reason for the double negative is cause the user doesn't actually elect the daily option (it's assigned by default and stays as long as they don't elect ClaimEarly)

I'm just going to rephrase it to "Confirm the user has not elected to claim early"